In [ ]:
def create_advanced_cnn_model():
    """
    İyileştirilmiş CNN modeli - Güçlü overfitting önleme
    
    Yeni Özellikler:
    - Daha güçlü Dropout oranları
    - Ek BatchNormalization katmanları
    - Daha fazla Regularization
    - İyileştirilmiş mimari
    """
    
    model = models.Sequential([
        # İlk Convolutional Blok
        layers.Conv2D(32, (3, 3), activation='relu', 
                     input_shape=(IMG_HEIGHT, IMG_WIDTH, 3),
                     kernel_regularizer=l2(0.002)),  # Artırıldı
        layers.BatchNormalization(),
        layers.Conv2D(32, (3, 3), activation='relu', 
                     kernel_regularizer=l2(0.002)),
        layers.MaxPooling2D((2, 2)),
        layers.Dropout(0.3),
        
        # İkinci Convolutional Blok
        layers.Conv2D(64, (3, 3), activation='relu', 
                     kernel_regularizer=l2(0.002)),
        layers.BatchNormalization(),
        layers.Conv2D(64, (3, 3), activation='relu', 
                     kernel_regularizer=l2(0.002)),
        layers.MaxPooling2D((2, 2)),
        layers.Dropout(0.35),  # Artırıldı
        
        # Üçüncü Convolutional Blok
        layers.Conv2D(128, (3, 3), activation='relu', 
                     kernel_regularizer=l2(0.002)),
        layers.BatchNormalization(),
        layers.Conv2D(128, (3, 3), activation='relu', 
                     kernel_regularizer=l2(0.002)),
        layers.MaxPooling2D((2, 2)),
        layers.Dropout(0.4),   # Artırıldı
        
        # Dördüncü Convolutional Blok
        layers.Conv2D(256, (3, 3), activation='relu', 
                     kernel_regularizer=l2(0.002)),
        layers.BatchNormalization(),
        layers.Dropout(0.4),
        
        # Global Average Pooling
        layers.GlobalAveragePooling2D(),
        
        # Dense Layers - Daha güçlü regularization
        layers.Dense(512, activation='relu', 
                    kernel_regularizer=l2(0.002)),
        layers.BatchNormalization(),
        layers.Dropout(0.6),    # Artırıldı
        
        layers.Dense(256, activation='relu', 
                    kernel_regularizer=l2(0.002)),
        layers.Dropout(0.5),    # Artırıldı
        
        # Output Layer
        layers.Dense(num_classes, activation='softmax')
    ])
    
    return model

# Model oluşturma
model = create_advanced_cnn_model()

print(" İyileştirilmiş Model Mimarisi Özeti:")
model.summary()

# Model karmaşıklığı analizi
total_params = model.count_params()
trainable_params = sum([tf.keras.backend.count_params(w) for w in model.trainable_weights])

print(f"\n Model Karmaşıklığı:")
print(f"    Toplam Parametre: {total_params:,}")
print(f"    Eğitilebilir Parametre: {trainable_params:,}")
print(f"    Tahmini Model Boyutu: {total_params * 4 / 1024 / 1024:.1f} MB")

I0000 00:00:1758905415.285109      19 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 13942 MB memory:  -> device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5
I0000 00:00:1758905415.285911      19 gpu_device.cc:2022] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 13942 MB memory:  -> device: 1, name: Tesla T4, pci bus id: 0000:00:05.0, compute capability: 7.5


 İyileştirilmiş Model Mimarisi Özeti:


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 222, 222, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 222, 222, 32)   │           128 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 220, 220, 32)   │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 110, 110, 32)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 110, 110, 32)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 108, 108, 64)   │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_1           │ (None, 108, 108, 64)   │           256 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_3 (Conv2D)               │ (None, 106, 106, 64)   │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 53, 53, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 53, 53, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_4 (Conv2D)               │ (None, 51, 51, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_2           │ (None, 51, 51, 128)    │           512 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 49, 49, 128)    │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 24, 24, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 24, 24, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_6 (Conv2D)               │ (None, 22, 22, 256)    │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_3           │ (None, 22, 22, 256)    │         1,024 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 22, 22, 256)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling2d        │ (None, 256)            │             0 │
│ (GlobalAveragePooling2D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 512)            │       131,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_4           │ (None, 512)            │         2,048 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 512)            │             

 Total params: 850,598 (3.24 MB)

 Trainable params: 848,614 (3.24 MB)

 Non-trainable params: 1,984 (7.75 KB)


 Model Karmaşıklığı:
    Toplam Parametre: 850,598
    Eğitilebilir Parametre: 848,614
    Tahmini Model Boyutu: 3.2 MB


Bu kısımda yaptığımız şey, derin öğrenme kullanarak görüntü sınıflandırma için **gelişmiş bir CNN (Convolutional Neural Network) modeli oluşturmak**. Modelin amacı, input olarak aldığı görsellerden özellikleri çıkarıp doğru sınıfı tahmin edebilmek.

Model, **Sequential yapı** ile katman katman inşa ediliyor. Öncelikle input boyutunu (`150,150,3`) olarak belirliyoruz; yani model 150x150 boyutunda ve 3 kanallı (RGB) görsellerle çalışacak. Çıkış katmanındaki nöron sayısı ise `num_classes` kadar, burada 6 sınıf için 6 nöron var.

Modelin gövdesi **4 adet Convolutional bloktan** oluşuyor. Her blok temel olarak şunları yapıyor:

Conv2D katmanı: Görsellerden kenar, renk, doku gibi özellikleri çıkarıyor. Blok ilerledikçe filtre sayısı artıyor (32 → 64 → 128 → 256), böylece daha karmaşık özellikler öğreniliyor. Aktivasyon olarak `ReLU` kullanılıyor; bu negatif değerleri sıfırlayarak öğrenmeyi hızlandırıyor ve nonlineer özellikler ekliyor. 

**BatchNormalization:** Katman çıktılarının dağılımını normalize ediyor. Bu sayede model daha stabil ve hızlı öğreniyor.

**MaxPooling2D:** Özellik haritasını küçülterek en baskın bilgiyi koruyor ve hesaplama maliyetini azaltıyor.

**Dropout:** Rastgele bazı nöronları kapatarak overfitting’i önlüyor.

Dördüncü blokta yalnızca Conv2D + BatchNorm + Dropout var çünkü artık yüksek seviyeli özellikler çıkarılmış ve boyutlar küçülmüş durumda.

Convolutional bloklardan sonra **fully connected katmanlar** geliyor:

**GlobalAveragePooling2D** → Son feature map’leri tek vektöre indiriyor. Bu, parametre sayısını azaltıyor ve overfitting riskini düşürüyor.

**Dense katmanları:** İlk dense katmanı 512 nöron, ikinci dense 256 nöron ve son dense çıkış için 6 nöron içeriyor. ReLU aktivasyonu öğrenme için kullanılırken, son katmanda `softmax` ile sınıflar arasında olasılık dağılımı elde ediliyor. Dropout ve BatchNormalization yine overfitting’i azaltmak ve öğrenmeyi stabilize etmek için eklenmiş. 

`model.summary()` çıktısı ise her katmanın tipini, çıktısının boyutunu ve parametre sayısını detaylı bir şekilde gösteriyor. Örneğin:

İlk Conv2D katmanı 32 filtre ile çalışıyor, çıktı boyutu (`148,148,32`) ve parametre sayısı 896.

MaxPooling2D ile boyut 73x73’e düşüyor, parametresiz.

Son dense katmanı 6 nöronlu ve softmax ile olasılık tahmini yapıyor.

Toplam parametre sayısı 850,598, bunun büyük kısmı trainable yani öğrenilebilir ağırlıklar, küçük bir kısmı non-trainable ve genellikle BatchNorm’dan geliyor.

Ayrıca TensorFlow GPU logları, modelin GPU üzerinde çalışacağını gösteriyor. GPU olmasa model CPU’da çalışacak ve çok daha yavaş eğitim olacaktı.

Özetle, bu kod parçası **görsellerden özellik çıkaran, fully connected katmanlarla sınıf tahmini yapan ve overfitting’i önleyici yöntemler içeren bir CNN modelini baştan sona oluşturuyor ve katmanların detaylarını gösteriyor.**


In [ ]:
class F1Score(keras.metrics.Metric):
    """Custom F1 Score metriği"""
    def __init__(self, name='f1_score', **kwargs):
        super().__init__(name=name, **kwargs)
        self.precision_metric = keras.metrics.Precision()
        self.recall_metric = keras.metrics.Recall()

    def update_state(self, y_true, y_pred, sample_weight=None):
        self.precision_metric.update_state(y_true, y_pred, sample_weight)
        self.recall_metric.update_state(y_true, y_pred, sample_weight)

    def result(self):
        precision = self.precision_metric.result()
        recall = self.recall_metric.result()
        return 2 * ((precision * recall) / (precision + recall + keras.backend.epsilon()))

    def reset_state(self):
        self.precision_metric.reset_state()
        self.recall_metric.reset_state()

# Model derleme - Daha düşük learning rate (DÜZELTILDI)
initial_lr = 0.0005  # Düşürüldü
model.compile(
    optimizer=Adam(learning_rate=initial_lr),
    loss='categorical_crossentropy',
    metrics=['accuracy', F1Score()]
)

print(f" Optimizer: Adam (lr={initial_lr}) - Düşürüldü")
print(f" Loss Function: Categorical Crossentropy")
print(f" Metrics: Accuracy, F1-Score")

# Güçlendirilmiş callback sistemi (DÜZELTILDI)
callbacks = [
    # Early Stopping - Daha sabırlı
    EarlyStopping(
        monitor='val_loss',
        patience=12,  # Artırıldı
        restore_best_weights=True,
        verbose=1,
        mode='min'
    ),
    
    # Learning Rate Reduction - Daha agresif
    ReduceLROnPlateau(
        monitor='val_loss',
        factor=0.2,   # Düşürüldü
        patience=5,   # Artırıldı
        min_lr=1e-8,  # Düşürüldü
        verbose=1,
        mode='min'
    ),
    
    # Model Checkpoint
    ModelCheckpoint(
        'best_intel_model.h5',
        monitor='val_f1_score',
        save_best_only=True,
        save_weights_only=False,
        mode='max',
        verbose=1
    )
]

print("\n İyileştirilmiş Callback Stratejileri:")
print("    EarlyStopping: val_loss, patience=12")
print("    ReduceLROnPlateau: factor=0.2, patience=5")
print("    ModelCheckpoint: en iyi F1-score modeli kaydet")

 Optimizer: Adam (lr=0.0005) - Düşürüldü
 Loss Function: Categorical Crossentropy
 Metrics: Accuracy, F1-Score

 İyileştirilmiş Callback Stratejileri:
    EarlyStopping: val_loss, patience=12
    ReduceLROnPlateau: factor=0.2, patience=5
    ModelCheckpoint: en iyi F1-score modeli kaydet


Bu kod, modelimizin eğitimini daha stabil ve performans odaklı hâle getirmek için yazıldı. Normal doğruluk yeterli olmadığında **F1-score** metriğini ekledik; böylece modelin tahminleri hem doğru hem dengeli şekilde ölçülüyor.

Learning rate’i **0.0005** olarak düşürdük, böylece ağırlıklar küçük ve kontrollü adımlarla güncelleniyor, eğitim sırasında ani değişimler ve kararsızlık önleniyor. Çok sınıflı sınıflandırma için categorical **crossentropy kayıp fonksiyonu** kullanıldı.

**Callback’ler** ise eğitimi akıllı hâle getiriyor:

**EarlyStopping** ile model artık iyileşmiyorsa duruyor ve en iyi ağırlıkları geri yüklüyor.

**ReduceLROnPlateau** kayıp takıldığında learning rate’i düşürerek modelin daha küçük adımlarla ilerlemesini sağlıyor.

**ModelCheckpoint** en iyi F1-score değerini yakaladığında modeli kaydediyor.

Kısaca, bu kod sayesinde model dengeli, güvenli ve kontrollü bir şekilde öğreniyor; overfitting riski azalıyor ve eğitim sırasında en iyi performansı veren model kaydediliyor.

In [ ]:
EPOCHS = 35  # Artırıldı, callbacks erken duracak

print(f" Epoch Sayısı: {EPOCHS}")
print(f" Tahmini Süre: {'20-40 dakika (GPU)' if len(tf.config.list_physical_devices('GPU')) > 0 else '3-6 saat (CPU)'}")

# Model eğitimi
print("\n İyileştirilmiş model eğitimi başlıyor...")

if data_paths and 'train_generator' in locals():
    # Gerçek veri ile eğitim
    history = model.fit(
        train_generator,
        epochs=EPOCHS,
        validation_data=validation_generator,
        callbacks=callbacks,
        verbose=1,
        steps_per_epoch=train_generator.samples // BATCH_SIZE,
        validation_steps=validation_generator.samples // BATCH_SIZE
    )
else:
    # Demo veri ile eğitim
    history = model.fit(
        X_train, y_train,
        batch_size=BATCH_SIZE,
        epochs=EPOCHS,
        validation_data=(X_val, y_val),
        callbacks=callbacks,
        verbose=1
    )

print(" Model eğitimi tamamlandı!")

 Epoch Sayısı: 35
 Tahmini Süre: 20-40 dakika (GPU)

 İyileştirilmiş model eğitimi başlıyor...
Epoch 1/35


I0000 00:00:1758905428.394030      86 service.cc:148] XLA service 0x7f4e1800c020 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1758905428.394773      86 service.cc:156]   StreamExecutor device (0): Tesla T4, Compute Capability 7.5
I0000 00:00:1758905428.394801      86 service.cc:156]   StreamExecutor device (1): Tesla T4, Compute Capability 7.5
I0000 00:00:1758905429.427706      86 cuda_dnn.cc:529] Loaded cuDNN version 90300


  1/350 ━━━━━━━━━━━━━━━━━━━━ 2:50:41 29s/step - accuracy: 0.1562 - f1_score: 0.0769 - loss: 5.6382

I0000 00:00:1758905447.682794      86 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


350/350 ━━━━━━━━━━━━━━━━━━━━ 0s 620ms/step - accuracy: 0.3717 - f1_score: 0.3444 - loss: 4.4470
Epoch 1: val_f1_score improved from -inf to 0.14846, saving model to best_intel_model.h5
350/350 ━━━━━━━━━━━━━━━━━━━━ 299s 771ms/step - accuracy: 0.3718 - f1_score: 0.3446 - loss: 4.4459 - val_accuracy: 0.1631 - val_f1_score: 0.1485 - val_loss: 6.0687 - learning_rate: 5.0000e-04
Epoch 2/35
  1/350 ━━━━━━━━━━━━━━━━━━━━ 1:01 176ms/step - accuracy: 0.4688 - f1_score: 0.5000 - loss: 3.5163
Epoch 2: val_f1_score did not improve from 0.14846
350/350 ━━━━━━━━━━━━━━━━━━━━ 40s 114ms/step - accuracy: 0.4688 - f1_score: 0.5000 - loss: 3.5163 - val_accuracy: 0.1620 - val_f1_score: 0.1366 - val_loss: 6.0500 - learning_rate: 5.0000e-04
Epoch 3/35
350/350 ━━━━━━━━━━━━━━━━━━━━ 0s 468ms/step - accuracy: 0.5251 - f1_score: 0.5045 - loss: 3.4136
Epoch 3: val_f1_score improved from 0.14846 to 0.42888, saving model to best_intel_model.h5
350/350 ━━━━━━━━━━━━━━━━━━━━ 204s 583ms/step - accuracy: 0.5252 - f1_score:

Bu kod, modelin eğitim sürecini yönetiyor ve eğitim sırasında ne olup bittiğini adım adım takip etmemizi sağlıyor.

Öncelikle **EPOCHS = 35** olarak ayarlandı. Bu, modelin veri üzerinden 35 kez geçeceği anlamına geliyor. Ancak callbacks sayesinde eğitim aslında daha erken durabilir; yani model yeterince öğrenip iyileşmeyince gereksiz yere uzun çalıştırılmıyor. Bu, hem zamanı hem de kaynakları koruyor.

Eğitim başladığında, model veriyi `train_generator` üzerinden alıyor. Her epoch’ta model ağırlıklarını güncelliyor, kaybı (loss) ve performans metriklerini (accuracy, F1-score) hesaplıyor. Örnek olarak:

İlk epoch’ta model çok düşük bir performansla başlıyor: accuracy ~0.35, F1-score ~0.32, loss ~4.62. Bu normal, çünkü ağırlıklar henüz rasgele başlıyor.

Epoch ilerledikçe model veriyi öğreniyor; accuracy ve F1-score yükseliyor, loss düşüyor. Örneğin 11. epoch’ta accuracy 0.70, F1-score 0.69, loss 1.85 civarında.

Callbacks burada kritik rol oynuyor:

**EarlyStopping:** Eğer val_loss uzun süre iyileşmezse eğitim duruyor. Bu, modelin overfitting yapmasını önlüyor.

**ReduceLROnPlateau:** Model belirli bir süre durursa learning rate düşüyor. Örneğin 28. epoch’ta learning rate 0.0005’den 0.0001’e düştü. Bu, modelin ağırlıkları daha küçük adımlarla güncellemesini sağlayarak daha stabil öğrenmesini sağlıyor.

**ModelCheckpoint:** Val_f1_score iyileştiğinde o epoch’taki modeli kaydediyor. Böylece eğitim sonunda en iyi performansa sahip model elimizde oluyor. Örneğin 23. epoch’ta val_f1_score 0.77257 oldu ve model kaydedildi; 31. epoch’ta ise 0.84031’e yükseldi ve yeniden kaydedildi.

Sonuçta, eğitim sonunda model `val_f1_score`’un en yüksek olduğu epoch’taki ağırlıkları geri yükledi. Bu, eğitimin sonunda en iyi performansı veren modelin elimizde kalmasını sağlıyor.

Özetle: bu kod **modeli kontrollü bir şekilde eğitiyor, performansı izliyor ve en iyi modelin kaydedilmesini sağlıyor**. Epoch’lar boyunca kayıp ve metrik değerleri bize modelin öğrenme sürecini low-level olarak gösteriyor: hangi aşamada hızla öğrendi, hangi aşamada durakladı ve learning rate’in değişimi ile performans nasıl optimize edildi.